src/dataset.cpp

```diff
@@ -574,8 +575,183 @@ Dataset concatenate(const Dataset &d1, const Dataset &d2, const Dim dim) {
   return out;
 }
 
+namespace neutron {
+namespace tof {
+Dataset tofToEnergy(const Dataset &d) {
+  // TODO Could in principle also support inelastic. Note that the conversion in
+  // Mantid is wrong since it handles inelastic data as if it were elastic.
+  if (d.contains(Coord::Ei{}) || d.contains(Coord::Ef{}))
+    throw std::runtime_error("Dataset contains Coord::Ei or Coord::Ef. "
+                             "However, conversion to Dim::Energy is currently "
+                             "only supported for elastic scattering.");
+
+  // 1. Compute conversion factor
+  const auto &compPos =
+      d.get<const Coord::ComponentInfo>()[0].get<const Coord::Position>();
+  // TODO Need a better mechanism to identify source and sample.
+  const auto &sourcePos = compPos[0];
+  const auto &samplePos = compPos[1];
+  const double l1 = (sourcePos - samplePos).norm();
+
+  const auto &dims = d.contains(Coord::Position{})
+                         ? d(Coord::Position{}).dimensions()
+                         : d(Coord::DetectorGrouping{}).dimensions();
+  Variable conversionFactor(Data::Value{}, dims);
+
+  MDZipView<const Coord::Position> specPos(d);
+  const auto factor = [&](const auto &item) {
+    const auto &pos = item.template get<Coord::Position>();
+    double l_total = l1 + (samplePos - pos).norm();
+    // TODO Up to physical constants.
+    return l_total * l_total;
+  };
+  // TODO Must also update unit of conversionFactor.
+  std::transform(specPos.begin(), specPos.end(),
+                 conversionFactor.get<Data::Value>().begin(), factor);
+
+  // 2. Transform variables
+  Dataset converted;
+  for (const auto &var : d) {
+    auto varDims = var.dimensions();
+    if (varDims.contains(Dim::Tof))
+      varDims.relabel(varDims.index(Dim::Tof), Dim::Energy);
+    if (var.tag() == Coord::Tof{}) {
```

Comment:
Are you following the approach currently taken by Mantid of having a "base" unit i.e. all units are converted to-from TOF? If not, then direct wavelength->energy conversions would be possible, and faster.